src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -124,19 +170,101 @@ private Set<TSource> FillSet()
                 return set;
             }
 
-            public TSource[] ToArray()
+            internal override IEnumerable<TSource> GetSourceEnumerable(int index)
             {
-                return FillSet().ToArray();
+                Debug.Assert(index == 0 || index == 1);
+                return index == 0 ? _first : _second;
             }
 
-            public List<TSource> ToList()
+            internal override UnionIterator<TSource> Union(IEnumerable<TSource> next)
             {
-                return FillSet().ToList();
+                return new UnionIteratorN<TSource>(this, next, 2);
             }
+        }
 
-            public int GetCount(bool onlyIfCheap)
+        private sealed class UnionIteratorN<TSource> : UnionIterator<TSource>
+        {
+            private readonly UnionIterator<TSource> _previous;
+            private readonly IEnumerable<TSource> _next;
+            private readonly int _sourceIndex;
```

Comment:
The point of calling this _nextIndex in Concat is that it was the index of _next, but here you've got _next and _sourceIndex.  We should be consistent... either _next and _nextIndex or _source and _sourceIndex.